File: src/http/jina_reranker.rs
use reqwest::blocking::Client;
use serde::{Deserialize, Serialize};

use crate::error::{MemlocalError, Result};
use crate::tools::RerankerProvider;

/// Jina AI reranker client using the public rerank API.
pub struct JinaReranker {
    api_key: String,
    model: String,
    base_url: String,
    client: Client,
}

impl JinaReranker {
    pub fn new(api_key: impl Into<String>) -> Self {
        Self {
            api_key: api_key.into(),
            model: "jina-reranker-v2-base-multilingual".to_string(),
            base_url: "https://api.jina.ai".to_string(),
            client: Client::new(),
        }
    }

    pub fn with_model(mut self, model: impl Into<String>) -> Self {
        self.model = model.into();
        self
    }

    pub fn with_base_url(mut self, base_url: impl Into<String>) -> Self {
        self.base_url = base_url.into();
        self
    }
}

impl RerankerProvider for JinaReranker {
    fn rerank(&self, query: &str, documents: &[String], top_k: usize) -> Result<Vec<(usize, f64)>> {
        if documents.is_empty() || top_k == 0 {
            return Ok(Vec::new());
        }

        let url = format!("{}/v1/rerank", self.base_url.trim_end_matches('/'));
        let body = RerankRequest {
            model: self.model.clone(),
            query: query.to_string(),
            documents: documents.to_vec(),
            top_n: top_k.min(documents.len()),
        };

        let response = self
            .client
            .post(&url)
            .header("Authorization", format!("Bearer {}", self.api_key))
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .map_err(|e| MemlocalError::Internal(format!("HTTP request failed: {e}")))?;

        if !response.status().is_success() {
            let status = response.status();
            let body_text = response
                .text()
                .unwrap_or_else(|_| "could not read body".to_string());
            return Err(MemlocalError::Internal(format!(
                "Jina API error ({status}): {body_text}"
            )));
        }

        let mut parsed: RerankResponse = response
            .json()
            .map_err(|e| MemlocalError::Internal(format!("Failed to parse response: {e}")))?;
        parsed
            .results
            .sort_by(|a, b| b.relevance_score.partial_cmp(&a.relevance_score).unwrap_or(std::cmp::Ordering::Equal));

        Ok(parsed
            .results
            .into_iter()
            .map(|result| (result.index, result.relevance_score))
            .collect())
    }
}

#[derive(Serialize)]
struct RerankRequest {
    model: String,
    query: String,
    documents: Vec<String>,
    top_n: usize,
}

#[derive(Deserialize)]
struct RerankResponse {
    results: Vec<RerankResult>,
}

#[derive(Deserialize)]
struct RerankResult {
    index: usize,
    relevance_score: f64,
}

File: flutter/lib/src/providers/jina_reranker.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'reranker_provider.dart';

/// [RerankerProvider] backed by the Jina AI rerank API (mirrors memlocal_core).
class JinaReranker implements RerankerProvider {
  /// Creates a reranker that calls the Jina AI rerank API with [apiKey], using
  /// [model]. [baseUrl] can point at a compatible endpoint.
  JinaReranker(
    this.apiKey, {
    this.model = 'jina-reranker-v2-base-multilingual',
    this.baseUrl = 'https://api.jina.ai/v1',
  });

  /// Jina API key sent as a bearer token. The caller owns this secret.
  final String apiKey;

  /// The Jina reranker model to use.
  final String model;

  /// Base URL of the (Jina-compatible) rerank API.
  final String baseUrl;

  @override
  Future<List<RerankResult>> rerank(
    String query,
    List<String> documents, {
    int topN = 5,
  }) async {
    if (documents.isEmpty) return const [];
    final res = await http.post(
      Uri.parse('$baseUrl/rerank'),
      headers: {
        'Authorization': 'Bearer $apiKey',
        'Content-Type': 'application/json',
      },
      body: jsonEncode({
        'model': model,
        'query': query,
        'documents': documents,
        'top_n': topN,
      }),
    );
    if (res.statusCode != 200) {
      throw Exception('Jina rerank ${res.statusCode}: ${res.body}');
    }
    final data = jsonDecode(res.body) as Map<String, dynamic>;
    final results = data['results'] as List;
    return results
        .map((r) => RerankResult(
              (r['index'] as num).toInt(),
              (r['relevance_score'] as num).toDouble(),
            ))
        .toList();
  }
}
